gnupg: change default keyserver to non-SKS

alyssais opened the pull request:
**I’ve asked the people behind keys.openpgp.org if they’re ready to become a distro default, and am awaiting a definitive answer from them. Until then this PR should not be merged.**

This is a breaking change, but I think it should still be backported to stable because the alternative is leaving people vulnerable.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

See https://gist.github.com/rjhansen/67ab921ffb4084c865b3618d6955275f.

The SKS network is vulnerable to certificate poisoning, which can destroy GnuPG installations. keys.openpgp.org is a new non-SKS keyserver that is resistant to this type of attack.

With such an attack being possible, it is unsafe to use SKS keyservers for almost anything, and so we should protect our users from a now unsafe default. keys.openpgp.org offers some (but not all) functionality of SKS, and is better than nothing.

This default is only present in `gnupg22`. `gnupg20` and `gnupg1orig` are not affected.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

alyssais committed:
gnupg: change default keyserver to non-SKS

See https://gist.github.com/rjhansen/67ab921ffb4084c865b3618d6955275f.

The SKS network is vulnerable to certificate poisoning, which can
destroy GnuPG installations. keys.openpgp.org is a new non-SKS keyserver
that is resistant to this type of attack.

With such an attack being possible, it is unsafe to use SKS keyservers
for almost anything, and so we should protect our users from a now
unsafe default. keys.openpgp.org offers some (but not all) functionality
of SKS, and is better than nothing.

This default is only present in gnupg22. gnupg20 and gnupg1orig are not
affected.

dkg commented:
hi!  as one of the folks whose OpenPGP certificate has been affected by this flooding attack, as the author of [draft-dkg-openpgp-abuse-resistant-keystore](https://tools.ietf.org/html/draft-dkg-openpgp-abuse-resistant-keystore) which describes the attack, and as one of the debian developers responsible for co-maintaining both GnuPG and SKS, i'm definitely interested in this sort of change.

I'm not convinced that the nice simple change you've done is correct, though. :/  What sort of tests have you run?

In particular, i'm concerned that `http_session_new()` in `dirmngr/http.c` adjusts what certificate authority to use based on whether the selected keyserver is the default keyserver or not.  if it is the default, it switches acceptable certificate authorities over to Kristian Fiskerstrand's custom authority, which is dedicated to providing certificates for the SKS pool.  That authority, of course, has *not* signed off on the X.509 cert offered by keys.openpgp.org.  So if it is willing to accept certifications from Kristian's CA to a server that is *not* the SKS pool, that seems like a problem.  I've [reported it upstream](https://dev.gnupg.org/T4593).

I note that later in the same function, the system CAs might be added as well, which might be why this appears to work for you.  i note that `add_system_cas`, but that will not be explicitly set to 1 when `is_hkps_pool` is set.  But perhaps it is coming in as 1 because the caller is setting `HTTP_FLAG_TRUST_SYS` unilaterally?  that would mean that dirmngr was being overly permissive in accepting system CAs for the HKPS SKS pool itself, when we know they should only be validated by Kristian's CA.  I've [reported that upstream as well](https://dev.gnupg.org/T4594).

lukateras reviewed (approved):
Looks good to me, except that this needs to include https://dev.gnupg.org/rG1c9cc97e9d47d73763810dcb4a36b6cdf31a2254 for the reasons @dkg has stated above.

alyssais committed:
gnupg: apply default server CA verification patch

See discussion at
https://github.com/NixOS/nixpkgs/pull/63952#issuecomment-507048690.

Upstream commit:

commit 1c9cc97e9d47d73763810dcb4a36b6cdf31a2254
Author: Daniel Kahn Gillmor <dkg@fifthhorseman.net>
Date:   Sun Jun 30 11:54:35 2019 -0400

    dirmngr: Only use SKS pool CA for SKS pool

    * dirmngr/http.c (http_session_new): when checking whether the
    keyserver is the HKPS pool, check specifically against the pool name,
    as ./configure might have been used to select a different default
    keyserver.  It makes no sense to apply Kristian's certificate
    authority to anything other than the literal host
    hkps.pool.sks-keyservers.net.

    Signed-off-by: Daniel Kahn Gillmor <dkg@fifthhorseman.net>
    GnuPG-Bug-Id: 4593

alyssais commented:
> hi!  as one of the folks whose OpenPGP certificate has been affected by this flooding attack, as the author of [draft-dkg-openpgp-abuse-resistant-keystore](https://tools.ietf.org/html/draft-dkg-openpgp-abuse-resistant-keystore) which describes the attack, and as one of the debian developers responsible for co-maintaining both GnuPG and SKS, i'm definitely interested in this sort of change.

Thanks very much for reviewing. It's great to get input from somebody
with so much upstream knowledge.

> I'm not convinced that the nice simple change you've done is correct, though. :/  What sort of tests have you run?

I note from the rest of your comment that you saw that it did in fact
appear to work, but just to be clear, I tested a --recv-keys and a
--refresh-keys.

alyssais commented:
Got the go ahead from the Hagrid (keys.openpgp.org) people, so I’m going to go ahead and merge this.

yegortimoshenko commented:
Thanks!
